### PR TITLE
[11.x] Add option to disable merging of base configuration

### PR DESCRIPTION
**Pull request by taka-oyama**
Addresses: https://github.com/laravel/framework/discussions/50876

## Context

https://github.com/laravel/framework/pull/47309 added defaults to various configurations which caused our tests to fail. 
For example, we have an `inspect` command which scans through and connects to external APIs such as `database.connections` and `queue.connections` to make sure that all connections are healthy. 
This change broke that and other similar commands that scan the configuration and connect to them. 

I've added `Illuminate\Foundation\Application::dontMergeBaseConfiguration()` which can be called during the bootstrap phase that will allow me to disable the merging of base configuration.

### CODE REVIEW

**Comment by taka-oyama** [comment]
It would probably look something like below.

```php
<?php

use Illuminate\Foundation\Application;
use Illuminate\Foundation\Configuration\Exceptions;
use Illuminate\Foundation\Configuration\Middleware;

$app = Application::configure(basePath: dirname(__DIR__))
    // ...
    ->withExceptions(function (Exceptions $exceptions) {
        //
    })->create();

$app->dontMergeBaseConfiguration()

return $app;
```

I guess I can clean this up a bit by adding the same method to `ApplicationBuilder`, but I use the `bootstrap/app.php` format from 10.x so I personally didn't need it.

File: src/Illuminate/Contracts/Foundation/Application.php [hunk 1]
```diff
@@ -237,6 +237,11 @@ public function setLocale($locale);
      */
     public function shouldSkipMiddleware();
 
+    /**
+     * @return bool
+     */
+    public function shouldMergeBaseConfiguration();
```
Comment (driesvints) [inline src/Illuminate/Contracts/Foundation/Application.php:243]: We can't add methods to a contract on a patch release.

Comment (taka-oyama) [reply src/Illuminate/Contracts/Foundation/Application.php]: I removed the method from interface and added `method_exists($app, 'shouldMergeBaseConfiguration')` check where the method is used. Would that work?

**Comment by taka-oyama** [comment]
Thanks for fixing the comments / formatting!

**Comment by jasonmccreary** [comment]
I try not to leave comments which may bring down the vibe of a PR. Generally speaking, I believe the framework should be flexible. Especially in its runtime configuration.

With that said, I would advise against using this method...

The merging of base configuration is what allows you to slim your Laravel application, as well as get new options for free. As someone who has facilitated over 100,000 Laravel application upgrades, I can say confidently this is a good thing. I have helped countless developers at their wits end wasting days on a random bug that turned out to be a single missed configuration option.

Using this method means you are fully responsible for maintaining all core configuration files in their entirety now and forever.

**Comment by taka-oyama** [comment]
> I would advise against using this method...

Do you have an alternate solution then?

**Comment by donnysim** [comment]
I think there's some missing points here for why some prefer to not merge it. As it stands now, it introduced a bunch of new environment variables, and as someone who upgraded apps back from 4.2, there's been a bunch of times that those variables were renamed. Having a static config and not mergeable removes the necessity to care about those changes on upgrade, which if you don't have that config "cloned" because you use that variable, you might not even know it changed if you don't read or missed it in the upgrade docs, and those included queue, mail, database variables. From me, that's a big no thank you, i prefer to have my variables static. If i need something new, i will put it in my config.

**Comment by jasonmccreary** [comment]
> but actually introducing problems

@mfn, can you share more detail on this?

**Comment by jasonmccreary** [comment]
@mfn, thanks. I still don't see the _problem_ - such as an error or incorrect overwrite.

What I see is more a desire for _full control_ over the configuration. As I noted in my original comment, that's fine when you assume full responsibility. You and other commenters here may truly understand what that means. I still don't think this method is needed and will likely be a footgun for most.

**Comment by donnysim** [comment]
The whole footgun is fake no matter if it's merged or not. if you don't override the config, any changes to config can be breaking, if you override the config with mergeable option, it will not merge anyway so you lose those new options. In any case you cannot depend on the new option being present, because even without the merge disable you are not guaranteed to have those new options.

**Comment by donnysim** [comment]
For example, take a clean laravel installation, it already comes with configs, now the framework decides to add, say `'lockout' => 60,` to the `auth.passwords.users` config, it will not be present for anyone unless you delete `auth.passwords.users` or `auth.passwords` or whole `auth` config file etc.. The whole concept that framework can magically add new options and it works for everyone is just not true. They might add root values, but even those risk clashing with existing user values and be different in structure etc. and even crash because of it. The whole config merging concept is just nice on paper, nothing more.

**Comment by jasonmccreary** [comment]
> The whole concept that framework can magically add new options and it works for everyone is just not true.

As someone who helped write this part, I have to say not true to your not true. 😉 

I do not wish to start a debate. Again, if this is something you all want to do, great. I left my disclaimer for everyone else.

**Comment by donnysim** [comment]
I tested it just to be doubly sure, so I'm 100% positive on what i said unless it's misunderstood, but yeah, no point in ranting on.

**Comment by pavloniym** [comment]
- I use L10
- I have `queue.connections` with default drivers published (!) in `configs/queue.php`
- I remove 'not-used' connections and add my custom connections with different redis dbs
- I have artisan command which iterate over `queue.connections` and clean all data using this connections
-  Everything is ok
- I upgraded to L11
- I have `queue.connections` with default drivers published (!) in `configs/queue.php`
- I remove 'not-used' connections and add my custom connections with different redis dbs
- I have artisan command which iterate over `queue.connections` and clean all data using this connections
- Why the hell framework merge default (not working and not existing) connections + my connections 
- Everything is broken in my pipelines

**Pull request merged**